content/guides/typescript-for-solid.mdx

```diff
@@ -0,0 +1,714 @@
+
+Consider a situation where you want to show a name only when it's defined using a Solid signal:
+
+```typescript
+const [name, setName] = createSignal<string>();
+return (
+  <Show when={(() => {
+      const v = data()
+      return v && v.type === "user" && v // T: User | undefined
+    })()}>
+  )
+```
+
+Accessing the callback may throw an error if the `when` prop is nullish or false at the time of access.
+Usually this is not the case, however in `onCleanup` or asynchronous contexts it can be.
+
+TypeScript may not like this.
+While you can understand that the second call to `name()` happens only if the first is truthy, TypeScript cannot. 
+It may flag an error indicating that `name()` could potentially be `undefined` when `.replace()` is called.
+
+There are a few ways to tackle this:
+
+1. **Non-Null Assertion**
+Use TypeScript's [non-null assertion operator (!).](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#non-null-assertion-operator-postfix-) 
+This operator tells TypeScript you're confident the value isn't `null` or `undefined`:
+
+```typescript
+return <Show when={name()}>Hello {name()!.replace(/\s+/g, "\xa0")}!</Show>;
+```
+While this may be a quick option, incorrectly asserting can lead to runtime errors if the value does turn out to be `null` or `undefined`.
+
+2. **Callback form of Show**
+
+Solid provides a callback form for its [`<Show>`](/reference/show) component. 
+In this form, the value that makes the when prop true is passed into the component:
+
+```typescript
+const [data, setData] = createSignal<User | Admin>() // T: Signal<User | Admin | undefined>
+
+return(
+    <Show when={(() => {
+      const v = data()
+      return v && v.type === "user" && v // T: User | undefined
+    })()}>
+  )
+```
+
+This approach takes advantage of the type information embedded in the [`<Show>`](/reference/show) component. 
+By temporarily storing the value `v`, you can use type narrowing to specify its type more accurately before returning it.
+
+You do have to be cautious, however, since accessing the callback may throw an error if the when prop is nullish or false at the time of access. 
+This is usually not an issue, but in asynchronous contexts or during [`onCleanup`](/reference/oncleanup), it may occur.
+
+<Aside>
+If you find the above method a bit shorthand, there's a more explicit alternative. 
+Solid's [`<Show>`](/reference/show) component actually performs memoization, so you don't necessarily need a memo for this:
+
+```tsx
+const user = createMemo(() => {
+  const v = data()
+  return v && v.type === "user" && v // Type: User | undefined
+})
+
+<Show when={user()}>
+  {user => <User user={user()} />}
+</Show>
+```
+While this approach is more explicit, it's worth noting that the primary example for the callback form should ideally not use this technique for the sake of clarity
+</Aside>
```

Comment:
I'm not sure how to get the code blocks to close properly.

```suggestion
## Control flow based narrowing

Control flow based narrowing refers to narrowing the type of a value via control flow. Consider the following:

```tsx
const user: User | undefined = maybeUser();
return <div>{user && user.name}</div>;
```

In Solid, however, Accessors cannot be narrowed in a similar way.

```tsx
const [user, setUser] = createSignal<User>();
return <div>{user() && user().name}</div>;
//                    ^ Object is possibly 'undefined'.
```

Similarly, with [`<Show>`](/reference/show), 

```tsx
return (
  <div>
    <Show when={user()}>
      {user().name /* Object is possibly 'undefined' */}
    </Show>
  </div>
);
```

In this case, using optional chaining is a good solution.

```tsx
return <div>{user()?.name}</div>;
```

And with `<Show>`,

```tsx
return (
  <div>
    <Show when={user()}>
      {nonNullishUser => nonNullishUser().name}
    </Show>
  </div>
);
```

This is similar to using the keyed option, except it provides an accessor instead so that
the children are not recreated whenever `when` changes. 

```tsx
return (
  <div>
    <Show keyed when={user()}>
      {nonNullishUser => nonNullishUser.name}
    </Show>
  </div>
);
```

Sometimes, this may not be possible. For instance, with a `UserPanel` component which strictly accepts a `User`:

```tsx
return <UserPanel user={user()} />;
//                        ^ Type 'undefined' is not assignable to type 'User'.
```

Consider refactoring `UserPanel` to accept `undefined`. Doing so would minimize the changes required when a `user` goes from `undefined` to `User`.

Otherwise, using `Show`'s callback form works,

```tsx
return (
  <Show when={user()}>
    {nonNullishUser => <UserPanel user={nonNullishUser()} />
  </Show>
);
```

<Aside>
Casting is also a valid solution, as long as the assumption is valid. For instance,

```tsx
return <div>{user() && (user() as User).name}</div>
```

However, runtime type errors may arise from doing so, notably when passing 
a type-cast value to a component, thereby discarding information that it may be nullish, 
then accessing it asynchronously (e.g. in an event handler or timeout) or in `onCleanup`.
</Aside>

`<Show>` only excludes `null`, `undefined`, and `false` from `when` when using the callback form. If you need to
discriminate a union, consider using a memo or computed signal as a workaround. 

```tsx
type User = Admin | OtherUser;
const admin = createMemo(() => {
  const u = user();
  return u && u.type === "admin" ? u : undefined;
});
return (
  <Show when={admin()}>
    {a => <AdminPanel user={a()} />}
  </Show>
);
```

<Aside>
The following alternative also works when using `Show`:

```tsx
<Show when={(() => {
  const u = user();
  return u && u.type === "admin" ? u : undefined;
})()}>
  {admin => <AdminPanel user={admin()} />}
</Show>
```
</Aside>
```